### PR TITLE
Icons management

### PR DESCRIPTION
**Pull request by thiloSchlemmer**
load icons-url from backend

### CODE REVIEW

File: src/services/IconsService.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+		return this._icons;
+	}
+
+	all() {
+		if (!this._icons) {
+			console.warn('IconsService not yet initialized');
+			return [];
+		}
+		return this._icons;
+	}
```
Comment (taulinger) [inline src/services/IconsService.js:34]: Caching for this service is good, but lazy loading should be sufficient here. We only load mandatory data eagerly (georesources, topics). This reduces the number of backend requests on startup.

**Comment by taulinger** [comment]
After thinking for a while, i would like to suggest creating a new module (iconSelect?) containing both the `ImageSelect` component and the `IconService` with its provider. 

Reason: It is not really a common reusable component (in the sense of the ba-button or other low-level components. How many `IconSelect` component instances will we use?) and we do not need the `IconService` on the global level also.

File: src/services/IconService.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+export class IconService {
+
+	constructor(iconProvider = loadBvvIcons, iconColorProvider = getBvvIconColor) {
+		this._iconProvider = iconProvider;
+		this._iconColorProvider = iconColorProvider;
+		this._default = this.createDefault();
+		this._icons = null;
+	}
+
+	createDefault() {
```
Comment (taulinger) [inline src/services/IconService.js:26]: Does this method have to be public?

File: src/services/IconService.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+	/**
+ *
+   * @param {string} url
+  */
+	decodeColor(url) {
+		return this._iconColorProvider(url);
+	}
+
+	isLocal(iconCandidate) {
```
Comment (taulinger) [inline src/services/IconService.js:95]: Does this method have to be public?

File: src/services/provider/iconUrl.provider.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  * A function that returns a url-string for the specified icon-id and color.
+  *
+  * @typedef {Function} iconUrlProvider
+  * @param {string} id the id of a specific icon
+  * @param {Array<number>} color the rgb-color
+  * @returns {string} the url to the icon
+  */
+
+/**
+ * creates a URL based on Image-Resources from BVV services
```
Comment (taulinger) [inline src/services/provider/iconUrl.provider.js:12]: [C]reates

File: test/service/IconService.test.js [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+			const icons = await instanceUnderTest.all();
+			const defaultIcon = instanceUnderTest.getDefault();
+
+			expect(icons.length).toBe(3);
+			expect(icons[0]).toEqual(defaultIcon);
+		});
+
+		it('initializes the service with default provider', async () => {
+			const instanceUnderTest = new IconService();
+			expect(instanceUnderTest._iconProvider).toEqual(loadBvvIcons);
```
Comment (taulinger) [inline test/service/IconService.test.js:51]: no test for default  `_iconColorProvider`

**Review by taulinger — COMMENTED** [review]
Looks good, great progress!
Let's talk about how to handle the "no-backend-available" case before merging.

File: src/services/IconService.js (hunk 3 above)
Comment (thiloSchlemmer) [reply src/services/IconService.js:95]: yes, it is used in olStyleUtils

File: test/service/IconService.test.js (hunk 5 above)
Comment (thiloSchlemmer) [reply test/service/IconService.test.js:51]: Thanks

File: src/services/IconService.js (hunk 3 above)
Comment (taulinger) [reply src/services/IconService.js:95]: ok

**Review by taulinger — APPROVED** [review]
Well done!

**Pull request merged**